### PR TITLE
Add ERC20Burnable preset

### LINKED ISSUE #237: Add ERC20 Burnable preset

### PR DESCRIPTION
**Pull request by koloz193**
Created an erc20 burnable preset from the mocked file and removed the mock since it's no longer needed.

Fixes #237 

Depends on:
- #240 
- #244

### CODE REVIEW

**Comment by andrew-fleming** [comment]
Hey @koloz193, thank you for submitting this PR! I know you're aware that we're considering some changes to the library (#267 and you mentioned in your comment^). Once we get this straightened out, we'll be able to review :)

**Comment by koloz193** [comment]
@andrew-fleming  makes sense. i ended up doing a fresh refactor based on the current state of main

**Comment by andrew-fleming** [comment]
@koloz193 ahh yeah, the rebasing conflicts must have looked like a nightmare

**Comment by koloz193** [comment]
> @koloz193 ahh yeah, the rebasing conflicts must have looked like a nightmare

yup, once i saw like 10 different files i quickly decided against trying to attempt the rebase

**Comment by andrew-fleming** [comment]
@koloz193 Ownable is merged! I appreciate your patience, sir 🙏 at your leisure, would you mind updating your branch?

**Comment by koloz193** [comment]
> 

awesome! will merge in main and clean it up!

**Review by andrew-fleming — COMMENTED** [review]
This looks great, sir! Thanks again for doing this :) I left a couple of comments

File: tests/token/erc20/test_ERC20_Burnable.py [hunk 1]
```diff
@@ -123,9 +130,63 @@ async def test_burn_from_zero_address(erc20_factory):
...
+    await signer.send_transaction(
+        account2, erc20.contract_address, 'burnFrom', [
+            account1.contract_address,
+            *AMOUNT
+        ])
+
+    new_balance = sub_uint(INIT_SUPPLY, AMOUNT)
+
+    execution_info = await erc20.balanceOf(account1.contract_address).invoke()
+    assert execution_info.result.balance == new_balance
```
[the remark this answers is not in the document]
Comment (koloz193) [reply tests/token/erc20/test_ERC20_Burnable.py:160]: good call, ill add these tests in

[the remark this answers is not in the document]
Comment (koloz193) [reply tests/token/erc20/test_ERC20_Burnable.py:171]: @andrew-fleming  added testing for the event. looks like `burnFrom` reverts when burn amount exceeds balance is covered by the following test and having account as the zero address results in an underflow error since the zero address cant give any approval

[the remark this answers is not in the document]
Comment (andrew-fleming) [reply tests/token/erc20/test_ERC20_Burnable.py:171]: Ah my mistake. Thank you for adding in the event test!

**Comment by koloz193** [comment]
@andrew-fleming @martriay any update on this 😄

**Comment by martriay** [comment]
This is currently on our [project roadmap](https://github.com/orgs/OpenZeppelin/projects/15), just that it's prioritized as "Low". This means we will tackle it whenever we finish higher priority issues, or we reprioritize based on new information. Sorry it's taking so long 🙇.

File: src/openzeppelin/token/erc20/presets/ERC20Burnable.cairo [hunk 2]
```diff
@@ -19,21 +21,33 @@ func constructor{
         symbol: felt,
         decimals: felt,
         initial_supply: Uint256,
-        recipient: felt
+        recipient: felt,
+        owner: felt
     ):
     ERC20.initializer(name, symbol, decimals)
     ERC20._mint(recipient, initial_supply)
+    Ownable.initializer(owner)
```
Comment (martriay) [inline src/openzeppelin/token/erc20/presets/ERC20Burnable.cairo:29]: why is it ownable?

File: src/openzeppelin/token/erc20/presets/ERC20Burnable.cairo [hunk 3]
```diff
@@ -150,11 +164,43 @@ end
...
+    return ()
+end
+
+@external
+func renounceOwnership{
+        syscall_ptr: felt*,
+        pedersen_ptr: HashBuiltin*,
+        range_check_ptr
+    }():
+    Ownable.renounce_ownership()
```
Comment (martriay) [inline src/openzeppelin/token/erc20/presets/ERC20Burnable.cairo:204]: I'm not sure I get why are these needed

File: tests/token/erc20/test_ERC20Burnable.py [hunk 4]
```diff
@@ -124,9 +131,74 @@ async def test_burn_from_zero_address(erc20_factory):
...
+    assert_event_emitted(
+        tx_exec_info,
+        from_address=erc20.contract_address,
+        name='Transfer',
+        data=[
+            account1.contract_address,
+            ZERO_ADDRESS,
+            *AMOUNT
+        ]
+    )
```
Comment (martriay) [inline tests/token/erc20/test_ERC20Burnable.py:167]: what do you think about splitting this into two different tests? that'd be consistent with `burn` whose storage effects and emitted events are tested separately

**Review by martriay — COMMENTED** [review]
@koloz193 sorry for taking so long! I updated the PR to match the latest directory structure (#350 ). I left a few comments but otherwise good to go!

**Comment by martriay** [comment]
Hey @koloz193! Do you think you will be able to finish this one off or should we pick it up? Thanks :)

**Comment by koloz193** [comment]
@martriay yup I'll wrap it up this weekend!

**Comment by martriay** [comment]
@koloz193 kind ping (:

File: src/openzeppelin/token/erc20/presets/ERC20Burnable.cairo (hunk 2 above)
Comment (koloz193) [reply src/openzeppelin/token/erc20/presets/ERC20Burnable.cairo:29]: good call, removing ownable

File: tests/token/erc20/test_ERC20Burnable.py (hunk 4 above)
Comment (koloz193) [reply tests/token/erc20/test_ERC20Burnable.py:167]: yea easy enough

**Comment by koloz193** [comment]
@martriay just pushed the changes, sorry for the delay but ran the tests and it looks good

File: src/openzeppelin/token/erc20/presets/ERC20Burnable.cairo [hunk 5]
```diff
@@ -1,11 +1,12 @@
 # SPDX-License-Identifier: MIT
+# OpenZeppelin Cairo Contracts v0.2.1 (token/erc20/presets/ERC20Burnable.cairo)
```
Comment (martriay) [inline src/openzeppelin/token/erc20/presets/ERC20Burnable.cairo:2]:
```suggestion
# OpenZeppelin Cairo Contracts v0.3.1 (token/erc20/presets/ERC20Burnable.cairo)
```
I think we need to update this

File: src/openzeppelin/token/erc20/presets/ERC20Burnable.cairo [hunk 6]
```diff
@@ -19,7 +20,8 @@ func constructor{
         symbol: felt,
         decimals: felt,
         initial_supply: Uint256,
-        recipient: felt
+        recipient: felt,
+        owner: felt
```
Comment (martriay) [inline src/openzeppelin/token/erc20/presets/ERC20Burnable.cairo:24]:
```suggestion
```
there's no owner anymore

File: tests/token/erc20/test_ERC20Burnable.py [hunk 7]
```diff
@@ -42,29 +46,32 @@ async def erc20_init(contract_classes):
             DECIMALS,
             *INIT_SUPPLY,
             account1.contract_address,        # recipient
+            account1.contract_address,        # owner
```
Comment (martriay) [inline tests/token/erc20/test_ERC20Burnable.py:49]:
```suggestion
```

**Review by martriay — CHANGES_REQUESTED** [review]
Looking good! Requested small changes, plus resolving merge conflicts I think we're good! Thanks again :)

**Comment by martriay** [comment]
Amazing! Now there's only the merge conflicts left.

**Comment by koloz193** [comment]
> Amazing! Now there's only the merge conflicts left.

not anymore 😄

**Review by martriay — APPROVED** [review]
